synapse/lib/ast.py

```diff
@@ -5408,48 +5398,99 @@ def allowed(x):
 
 class EditTagAdd(Edit):
 
+    def __init__(self, astinfo, kids=(), istry=False):
+        Edit.__init__(self, astinfo, kids=kids)
+        self.excignore = ()
+        if istry:
+            self.excignore = (s_exc.BadTypeValu,)
+
     async def run(self, runt, genr):
 
         self.reqNotReadOnly(runt)
 
-        if len(self.kids) > 1 and isinstance(self.kids[0], Const) and (await self.kids[0].compute(runt, None)) == '?':
-            oper_offset = 1
-        else:
-            oper_offset = 0
-
-        excignore = (s_exc.BadTypeValu,) if oper_offset == 1 else ()
-
-        hasval = len(self.kids) > 2 + oper_offset
+        namekid = self.kids[0]
 
         valu = (None, None)
+        valukid = None
+        if len(self.kids) == 3:
+            valukid = self.kids[2]
 
         async for node, path in genr:
 
             try:
-                names = await self.kids[oper_offset].computeTagArray(runt, path, excignore=excignore)
-            except excignore:
+                names = await namekid.computeTagArray(runt, path, excignore=self.excignore)
+            except self.excignore:
                 yield node, path
                 await asyncio.sleep(0)
                 continue
 
+            if valukid is not None:
+                valu = await valukid.compute(runt, path)
+                valu = await s_stormtypes.toprim(valu)
+
             for name in names:
 
                 try:
                     parts = name.split('.')
-
                     runt.layerConfirm(('node', 'tag', 'add', *parts))
 
-                    if hasval:
-                        valu = await self.kids[2 + oper_offset].compute(runt, path)
-                        valu = await s_stormtypes.toprim(valu)
                     await node.addTag(name, valu=valu)
-                except excignore:
+                except self.excignore:
```

Comment:
while we're in here, does this effective `continue` need a `sleep(0)`?

Reply:
yeah might as well add one